x-pack/plugins/lens/public/heatmap_visualization/chart_component.tsx

```diff
@@ -0,0 +1,306 @@
+import { LensBrushEvent, LensFilterEvent } from '../types';
+import { desanitizeFilterContext } from '../utils';
+import { search } from '../../../../../src/plugins/data/public';
+import { EmptyPlaceholder } from '../shared_components';
+import { LensIconChartHeatmap } from '../assets/chart_heatmap';
+
+declare global {
+  interface Window {
+    /**
+     * Flag used to enable debugState on elastic charts
+     */
+    _echDebugStateFlag?: boolean;
+  }
+}
+
+export const HeatmapComponent: FC<HeatmapRenderProps> = ({
+  data,
+  args,
+  timeZone,
+  formatFactory,
+  chartsThemeService,
+  onClickValue,
+  onSelectRange,
+}) => {
+  const chartTheme = chartsThemeService.useChartsTheme();
+  const isDarkTheme = chartsThemeService.useDarkMode();
+
+  const tableId = Object.keys(data.tables)[0];
+  const table = data.tables[tableId];
+
+  const xAxisColumnIndex = table.columns.findIndex((v) => v.id === args.xAccessor);
+  const yAxisColumnIndex = table.columns.findIndex((v) => v.id === args.yAccessor);
+
+  const xAxisColumn = table.columns[xAxisColumnIndex];
+  const yAxisColumn = table.columns[yAxisColumnIndex];
+  const valueColumn = table.columns.find((v) => v.id === args.valueAccessor);
+
+  if (!xAxisColumn || !valueColumn) {
+    // Chart is not ready
+    return null;
+  }
+
+  let chartData = table.rows.filter((v) => typeof v[args.valueAccessor!] === 'number');
+
+  if (!yAxisColumn) {
+    // required for tooltip
+    chartData = chartData.map((row) => {
+      return {
+        ...row,
+        unifiedY: '',
+      };
+    });
+  }
+
+  const xAxisMeta = xAxisColumn.meta;
+  const isTimeBasedSwimLane = xAxisMeta.type === 'date';
+
+  // Fallback to the ordinal scale type when a single row of data is provided.
+  // Related issue https://github.com/elastic/elastic-charts/issues/1184
+  const xScaleType =
+    isTimeBasedSwimLane && chartData.length > 1 ? ScaleType.Time : ScaleType.Ordinal;
+
+  const xValuesFormatter = formatFactory(xAxisMeta.params);
+  const valueFormatter = formatFactory(valueColumn.meta.params);
+
+  // Enable xDomain when https://github.com/elastic/elastic-charts/issues/1165 is resolved.
+  // @ts-ignore
+  const xDomain = (() => {
+    if (!isTimeBasedSwimLane) return null;
+    const dateInterval = search.aggs.getDateHistogramMetaDataByDatatableColumn(xAxisColumn)
+      ?.interval;
+    if (!dateInterval) return null;
+    const intervalDuration = search.aggs.parseInterval(dateInterval);
+    if (!intervalDuration) return null;
+    const minInterval = intervalDuration.as('milliseconds');
+    const dateRangeMin = data.dateRange?.fromDate.getTime();
+
+    if (!dateRangeMin) {
+      return;
+    }
+
+    const actualMin = dateRangeMin - (dateRangeMin % minInterval);
+
+    return {
+      min: actualMin,
+      max: data.dateRange?.toDate.getTime(),
+      minInterval,
+    };
+  })();
+
+  // @ts-ignore
+  const onElementClick: ElementClickListener = (e: HeatmapElementEvent[]) => {
+    const cell = e[0][0];
```

Comment:
nit: Please cast the `(e: HeatmapElementEvent[]) => void` to `ElementClickListener` instead of ignoring the ts error. That's retains at least some type safety.

Reply:
Fixed in https://github.com/elastic/kibana/pull/97978/commits/ef9547158899434cd3ad7f0a15c1ca8543f61a8c